processor/src/tests/u32_ops.rs

```diff
@@ -0,0 +1,950 @@
+        } else {
+            // should execute left shift
+            let result =  (a << b) % U32_BOUND;
+            test_execution(&asm_opcode, &[a], &[result])
+        }
+    }
+
+    #[test]
+    // https://github.com/maticnetwork/miden/issues/76
+    fn u32shr_proptest(a in any::<u64>(), b in 0_u32..32) {
+        let asm_opcode = format!("u32shr.{}", b);
+
+        if a >= U32_BOUND {
+            // should fail
+            test_execution_failure(&asm_opcode, &[a], "FailedAssertion");
+        } else {
+            // should execute right shift
+            let result =  a >> b;
+            test_execution(&asm_opcode, &[a], &[result])
+        }
+    }
+
+    #[test]
+    // https://github.com/maticnetwork/miden/issues/76
+    fn u32rotl_proptest(a in any::<u64>(), b in 0_u32..32) {
+        let op_base = "u32rotl";
+        let asm_opcode = format!("{}.{}", op_base, b);
+
+        if a >= U32_BOUND {
+            // should fail
+            test_execution_failure(&asm_opcode, &[a], "FailedAssertion");
+        } else {
+            // should execute left bit rotation
+            test_execution(&asm_opcode, &[a], &[(a as u32).rotate_left(b) as u64])
+        }
+    }
+
+    #[test]
+    // https://github.com/maticnetwork/miden/issues/76
+    fn u32rotr_proptest(a in any::<u64>(), b in 0_u32..32) {
+        let op_base = "u32rotr";
+        let asm_opcode = format!("{}.{}", op_base, b);
+
+        if a >= U32_BOUND {
+            // should fail
+            test_execution_failure(&asm_opcode, &[a], "FailedAssertion");
+        } else {
+            // should execute right bit rotation
+            test_execution(&asm_opcode, &[a], &[(a as u32).rotate_right(b) as u64])
+        }
+    }
+}
+
+// U32 OPERATIONS TESTS - RANDOMIZED - COMPARISON OPERATIONS
+// ================================================================================================
+
+proptest! {
+    #[test]
+    #[ignore = "unimplemented"]
+    fn u32eq_proptest(a in any::<u64>(), b in any::<u64>()) {
+        let asm_opcode = "u32and";
+        let values = [b, a];
+
+        if a >= U32_BOUND {
+            // should fail when a is out of bounds
+            test_execution_failure(asm_opcode, &values, "FailedAssertion");
+        } else if b >= U32_BOUND {
+            // should fail when b is out of bounds and provided as a parameter
+            test_param_out_of_bounds(asm_opcode, b);
+            // should fail when b is out of bounds and provided via the stack
+            test_execution_failure(asm_opcode, &values, "FailedAssertion");
+        } else {
+            // should test for equality
+            let expected = if a==b { 1 } else { 0 };
+            // b provided via the stack
+            test_execution(asm_opcode, &values, &[expected]);
+            // b provided as a parameter
+            test_execution(format!("{}.{}", asm_opcode, b).as_str(), &[a], &[expected]);
+        }
```

Comment:
I'm curious if proptest hits this cases regularly. If it sampled u64 values uniformly at random, these would be hit very rarely. But I believe it is actually smarter than that. So, just want to confirm.

Reply:
I wanted to test it with a couple runs, but this test is currently failing because the comparison ops aren't implemented to spec yet.

However - I'm not sure how much smarter proptest actually is - it's great for finding the minimal failing input once it's got an error, but for the initial values I think it may just be sampling uniformly at random as you said. I looked into this a bit with the last PR that I used proptest on, because it was missing an important edge case and I was wondering if a check I was doing inside the proptest was even needed. I ended up implementing the edge case manually, but I left the check inside the proptest because I didn't want a false positive.

From their crate:

>Given infinite time, property testing will eventually explore the whole input space to a test. However, time is not infinite, so only a randomly sampled portion of the input space can be explored. This means that property testing is extremely unlikely to find single-value edge cases in a large space.

Reply:
Should we then sample from a `u32` space for the "else" clause?

Reply:
Ahh fair point. In that case, is it just better to do 2 tests - one that samples u32 for the valid case and one that samples u64 for failing & maybe even filters out the u32s? (Although it seems like filter may not be worth the cost. Opinion on that?)

Reply:
Maybe invalid cases can be handled in manual tests? And the test validity of operations over valid inputs using randomized tests?